File: app/src/main/java/no/nav/tilbakemeldingsmottak/bigquery/serviceklager/ServiceklagerBigQuery.java
package no.nav.tilbakemeldingsmottak.bigquery.serviceklager;

import com.google.api.client.util.DateTime;
import com.google.cloud.bigquery.BigQuery;
import com.google.cloud.bigquery.InsertAllRequest;
import com.google.cloud.bigquery.TableId;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import no.nav.tilbakemeldingsmottak.serviceklage.Serviceklage;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.env.Environment;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Arrays;
import java.util.HashMap;

@Component
@RequiredArgsConstructor
@Slf4j
public class ServiceklagerBigQuery {

    public static final String TABLE_NAME = "serviceklager";
    private final Environment env;
    private final BigQuery bigQueryClient;
    @Value("${big_query_dataset}")
    private String dataset;

    private DateTime getDateTime(LocalDateTime localDateTime) {
        if (localDateTime == null) {
            return null;
        }
        var epoch = localDateTime.atZone(ZoneId.of("Europe/Oslo")).toInstant().toEpochMilli();
        return new DateTime(epoch);
    }

    public void insertServiceklage(Serviceklage serviceklage, ServiceklageEventTypeEnum eventType) {

        try {
            var map = new HashMap<String, Object>();
            map.put("event_type", eventType.toString());
            map.put("serviceklage_id", serviceklage.getServiceklageId());
            map.put("journalpost_id", serviceklage.getJournalpostId());
            map.put("opprettet_dato", getDateTime(serviceklage.getOpprettetDato()));
            map.put("klagen_gjelder_id", serviceklage.getKlagenGjelderId());
            map.put("klagetyper", serviceklage.getKlagetyper());
            map.put("gjelder_sosialhjelp", serviceklage.getGjelderSosialhjelp());
            map.put("klagetekst", serviceklage.getKlagetekst());
            map.put("behandles_som_serviceklage", serviceklage.getBehandlesSomServiceklage());
            map.put("behandles_som_serviceklage_utdypning", serviceklage.getBehandlesSomServiceklageUtdypning());
            map.put("fremmet_dato", getDateTime(serviceklage.getFremmetDato().atStartOfDay()));
            map.put("innsender", serviceklage.getInnsender());
            map.put("kanal", serviceklage.getKanal());
            map.put("kanal_utdypning", serviceklage.getKanalUtdypning());
            map.put("enhetsnummer_paaklaget", serviceklage.getEnhetsnummerPaaklaget());
            map.put("enhetsnummer_behandlende", serviceklage.getEnhetsnummerBehandlende());
            map.put("gjelder", serviceklage.getGjelder());
            map.put("beskrivelse", serviceklage.getBeskrivelse());
            map.put("ytelse", serviceklage.getYtelse());
            map.put("tema_utdypning", serviceklage.getTemaUtdypning());
            map.put("veiledning", serviceklage.getVeiledning());
            map.put("utfall", serviceklage.getUtfall());
            map.put("aarsak", serviceklage.getAarsak());
            map.put("tiltak", serviceklage.getTiltak());
            map.put("svarmetode", serviceklage.getSvarmetode());
            map.put("svarmetode_utdypning", serviceklage.getSvarmetodeUtdypning());
            map.put("avsluttet_dato", getDateTime(serviceklage.getAvsluttetDato()));
            map.put("skjema_versjon", serviceklage.getSkjemaVersjon());
            map.put("relatert", serviceklage.getRelatert());
            map.put("klagetype_utdypning", serviceklage.getKlagetypeUtdypning());
            map.put("innlogget", serviceklage.getInnlogget());

            InsertAllRequest request = InsertAllRequest.newBuilder(TableId.of(dataset, TABLE_NAME)).addRow(map).build();

            log.info("Inserting rows into Big Query");

            if (Arrays.asList(env.getActiveProfiles()).contains("local")) {
                log.info("Skal ikke legge til rader i big query i lokal env");
                return;
            }

            var insertAllResponse = bigQueryClient.insertAll(request);
            if (insertAllResponse.hasErrors()) {
                insertAllResponse.getInsertErrors().forEach((k, v) -> log.error("Error inserting row: " + k + " " + v.toString()));
            }
        } catch (Exception e) {
            log.error("Error inserting into BigQuery", e);
        }
    }
}

File: app/src/main/kotlin/no/nav/tilbakemeldingsmottak/bigquery/serviceklager/ServiceklagerBigQuery.kt
package no.nav.tilbakemeldingsmottak.bigquery.serviceklager

import com.google.api.client.util.DateTime
import com.google.cloud.bigquery.BigQuery
import com.google.cloud.bigquery.BigQueryError
import com.google.cloud.bigquery.InsertAllRequest
import com.google.cloud.bigquery.TableId
import no.nav.tilbakemeldingsmottak.domain.models.Serviceklage
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.core.env.Environment
import org.springframework.stereotype.Component
import java.time.LocalDateTime
import java.time.ZoneId

@Component
class ServiceklagerBigQuery(private val env: Environment, private val bigQueryClient: BigQuery) {

    companion object {
        const val TABLE_NAME = "serviceklager"
    }

    @Value("\${big_query_dataset}")
    private val dataset: String? = null

    private val logger = LoggerFactory.getLogger(javaClass)

    private fun getDateTime(localDateTime: LocalDateTime?): DateTime? {
        if (localDateTime == null) {
            return null
        }
        val epoch = localDateTime.atZone(ZoneId.of("Europe/Oslo")).toInstant().toEpochMilli()
        return DateTime(epoch)
    }

    fun insertServiceklage(serviceklage: Serviceklage, eventType: ServiceklageEventType) {
        try {
            val map = HashMap<String, Any?>()
            map["event_type"] = eventType.name
            map["serviceklage_id"] = serviceklage.serviceklageId
            map["journalpost_id"] = serviceklage.journalpostId
            map["opprettet_dato"] = getDateTime(serviceklage.opprettetDato)
            map["klagen_gjelder_id"] = serviceklage.klagenGjelderId
            map["klagetyper"] = serviceklage.klagetyper
            map["gjelder_sosialhjelp"] = serviceklage.gjelderSosialhjelp
            map["klagetekst"] = null
            map["behandles_som_serviceklage"] = serviceklage.behandlesSomServiceklage
            map["behandles_som_serviceklage_utdypning"] = serviceklage.behandlesSomServiceklageUtdypning
            map["fremmet_dato"] =
                getDateTime(if (serviceklage.fremmetDato != null) serviceklage.fremmetDato?.atStartOfDay() else null)
            map["innsender"] = serviceklage.innsender
            map["kanal"] = serviceklage.kanal
            map["kanal_utdypning"] = serviceklage.kanalUtdypning
            map["enhetsnummer_paaklaget"] = serviceklage.enhetsnummerPaaklaget
            map["enhetsnummer_behandlende"] = serviceklage.enhetsnummerBehandlende
            map["gjelder"] = serviceklage.gjelder
            map["beskrivelse"] = serviceklage.beskrivelse
            map["ytelse"] = serviceklage.ytelse
            map["tema"] = serviceklage.tema
            map["tema_utdypning"] = serviceklage.temaUtdypning
            map["veiledning"] = serviceklage.veiledning
            map["utfall"] = serviceklage.utfall
            map["aarsak"] = serviceklage.aarsak
            map["tiltak"] = serviceklage.tiltak
            map["svarmetode"] = serviceklage.svarmetode
            map["svarmetode_utdypning"] = serviceklage.svarmetodeUtdypning
            map["avsluttet_dato"] = getDateTime(serviceklage.avsluttetDato)
            map["skjema_versjon"] = serviceklage.skjemaVersjon
            map["klassifisering_json"] = serviceklage.klassifiseringJson
            map["relatert"] = serviceklage.relatert
            map["klagetype_utdypning"] = serviceklage.klagetypeUtdypning
            map["innlogget"] = serviceklage.innlogget
            map["oppgave_id"] = serviceklage.oppgaveId

            val request = InsertAllRequest.newBuilder(TableId.of(dataset, TABLE_NAME)).addRow(map).build()
            logger.info("Inserting rows into Big Query")

            if (env.activeProfiles.contains("local") || env.activeProfiles.contains("itest")) {
                logger.info("Skal ikke legge til rader i big query i lokal env")
                return
            }

            val insertAllResponse = bigQueryClient.insertAll(request)
            if (insertAllResponse.hasErrors()) {
                insertAllResponse.insertErrors.forEach { (k: Long, v: List<BigQueryError?>) ->
                    logger.error("Error inserting row: $k $v")
                }
            }
        } catch (e: Exception) {
            logger.error("Error inserting into BigQuery", e)
        }
    }
}
